waspc/data/Generator/templates/sdk/wasp/client/operations/queries/core.ts

```diff
@@ -1,53 +1,64 @@
 import { Route } from 'wasp/client'
-import type { Expand, _Awaited, _ReturnType } from 'wasp/universal/types'
-import { type Query } from '../core.js'
+import type { _Awaited, _ReturnType } from 'wasp/universal/types'
+import type { 
+  ClientOperation,
+  GenericBackendOperation,
+  OperationRpcFor,
+  QueryFunction,
+  QueryMetadata,
+} from '../rpc.js'
 import { callOperation, makeOperationRoute } from '../internal/index.js'
 import {
   addResourcesUsedByQuery,
   getActiveOptimisticUpdates,
 } from '../internal/resources'
 
-export function createQuery<BackendQuery extends GenericBackendQuery>(
+// PRIVATE API (unsed in SDK)
+export function createQuery<BackendQuery extends GenericBackendOperation>(
   relativeQueryPath: string,
   entitiesUsed: string[]
 ): QueryFor<BackendQuery> {
   const queryRoute = makeOperationRoute(relativeQueryPath)
+  const queryCacheKey = [relativeQueryPath]
 
-  async function query(queryKey, queryArgs) {
+  const queryFn: QueryFunctionFor<BackendQuery> = async (queryArgs) => { 
     const serverResult = await callOperation(queryRoute, queryArgs)
-    return getActiveOptimisticUpdates(queryKey).reduce(
+    return getActiveOptimisticUpdates(queryCacheKey).reduce(
       (result, update) => update(result),
       serverResult,
     )
   }
 
-  addMetadataToQuery(query, { relativeQueryPath, queryRoute, entitiesUsed })
-
-  return query
+  return buildAndRegisterQuery(
+    queryFn,
+    { queryCacheKey, queryRoute, entitiesUsed },
+  )
 }
 
-// PRIVATE API
-export function addMetadataToQuery(
-  query: (...args: any[]) => Promise<unknown>,
-  metadata: {
-    relativeQueryPath: string
-    queryRoute: Route
-    entitiesUsed: string[]
-  }
-): void
-
-// PRIVATE API
-export function addMetadataToQuery(
-  query,
-  { relativeQueryPath, queryRoute, entitiesUsed }
-) {
-  query.queryCacheKey = [relativeQueryPath]
+// PRIVATE API (used in SDK)
+export function buildAndRegisterQuery<Input, Output>(
+  queryFn: ClientOperation<Input, Output>,
+  { queryCacheKey, queryRoute, entitiesUsed }: 
+  { queryCacheKey: string[], queryRoute: Route, entitiesUsed: string[] }
+): QueryForFunction<typeof queryFn> {
+  const query = queryFn as QueryForFunction<typeof queryFn>
+
+  query.queryCacheKey = queryCacheKey 
   query.route = queryRoute
   addResourcesUsedByQuery(query.queryCacheKey, entitiesUsed)
+
+  return query
+
 }
 
-export type QueryFor<BackendQuery extends GenericBackendQuery> = 
-  Query<Parameters<BackendQuery>[0], _Awaited<_ReturnType<BackendQuery>>>
+// PRIVATE API (but should maybe be public, users define values of this type)
+export type QueryFor<BackendQuery extends GenericBackendOperation> =
+  QueryForFunction<QueryFunctionFor<BackendQuery>>
+
+type QueryFunctionFor<BackendQuery extends GenericBackendOperation> =
+  OperationRpcFor<BackendQuery>
 
+// PRIVATE API (needed in SDK)
+type QueryForFunction<QF extends QueryFunction<never, unknown>> = 
+  QF & QueryMetadata
```

Comment:
I found this block hard to parse, maybe put some _why_ comment on top to make it easier for us in the future to get back into this code.